File: apps/fin_core/fin_core/accounting/report/analytics_summary/analytics_summary.py
import frappe
from frappe import _
from frappe.utils import flt


def execute(filters=None):
    if not filters: filters = {}

    columns = get_columns()
    data = get_data(filters)

    return columns, data


def get_columns():
    return [
        {
            "label": _("Аналітика"),
            "fieldname": "analytics",
            "fieldtype": "Link",
            "options": "Analytics", # Це критично для відображення як посилання
            "width": 160
        },
        {"label": _("Поч. залишок"), "fieldname": "opening_balance", "fieldtype": "Currency", "width": 120},
        {"label": _("Прихід (+)"), "fieldname": "inbound", "fieldtype": "Currency", "width": 120},
        {"label": _("Витрата (-)"), "fieldname": "outbound", "fieldtype": "Currency", "width": 120},
        {"label": _("Кінц. залишок"), "fieldname": "closing_balance", "fieldtype": "Currency", "width": 140},
    ]


def get_data(filters):
    from_date = filters.get("from_date")
    to_date = filters.get("to_date")

    # Отримуємо всі аналітики
    analytics_list = frappe.get_all("Analytics", fields=["name"])
    report_data = []

    for a in analytics_list:
        name = a.name

        # 1. Рахуємо вхідний залишок (сума ТО - сума FROM до from_date)
        opening = frappe.db.sql("""
            SELECT 
                SUM(CASE WHEN to_analytics = %s THEN amount ELSE 0 END) -
                SUM(CASE WHEN from_analytics = %s THEN amount ELSE 0 END) as balance
            FROM `tabFin Transaction`
            WHERE (to_analytics = %s OR from_analytics = %s)
              AND docstatus = 1
              AND date < %s
        """, (name, name, name, name, from_date), as_dict=1)[0].balance or 0.0

        # 2. Рахуємо обороти за період
        period_data = frappe.db.sql("""
            SELECT 
                SUM(CASE WHEN to_analytics = %s THEN amount ELSE 0 END) as inbound,
                SUM(CASE WHEN from_analytics = %s THEN amount ELSE 0 END) as outbound
            FROM `tabFin Transaction`
            WHERE docstatus = 1
              AND date BETWEEN %s AND %s
        """, (name, name, from_date, to_date), as_dict=1)[0]

        inbound = period_data.inbound or 0.0
        outbound = period_data.outbound or 0.0
        closing = opening + inbound - outbound

        if opening != 0 or inbound != 0 or outbound != 0:
            report_data.append({
                "analytics": name,
                "opening_balance": opening,
                "inbound": inbound,
                "outbound": outbound,
                "closing_balance": closing
            })

    return report_data

File: apps/fin_core/fin_core/accounting/report/analytics_summary/analytics_summary.js
frappe.query_reports["Analytics Summary"] = {
    "filters": [
        {
            "fieldname": "from_date",
            "label": __("З дати"),
            "fieldtype": "Date",
            "default": frappe.datetime.month_start(),
            "reqd": 1
        },
        {
            "fieldname": "to_date",
            "label": __("По дату"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "reqd": 1
        }
    ],

    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        if (column.fieldname == "closing_balance" && data && data.closing_balance < 0) {
            value = `<span style="color:red; font-weight:bold;">${value}</span>`;
        }
        return value;
    },

    // ВАЖЛИВО: Переносимо сюди для гарантованого спрацювання
    "get_datatable_options": function(options) {
        return Object.assign(options, {
            onClick: ({ column, row }) => {
                if (!row) return;

                const data = row.dict; // В нових версіях DataTable дані лежать в .dict
                const fieldname = column.id || column.fieldname;

                if (fieldname === "analytics" && data && data.analytics) {
                    // Формуємо прості фільтри для надійності
                    frappe.set_route("List", "Fin Transaction", {
                        "from_analytics": data.analytics,
                        "date": [">=", frappe.query_report.get_filter_value("from_date")],
                        // Ми відкриваємо загальний список, зазвичай по одній стороні транзакції
                        // Якщо треба "OR", краще відкривати спеціальний звіт Ledger
                    });
                }
            }
        });
    },

};